src/Microsoft.OData.Core/Json/PooledByteBufferWriter.cs

```diff
@@ -0,0 +1,221 @@
+        /// </summary>
+        public int WrittenCount
+        {
+            get
+            {
+                Debug.Assert(rentedBuffer != null);
+                return this.index;
+            }
+        }
+
+        /// <summary>
+        /// Gets the total amount of space within the underlying buffer.
+        /// </summary>
+        public int Capacity
+        {
+            get
+            {
+                Debug.Assert(this.rentedBuffer != null);
+                return this.rentedBuffer.Length;
+            }
+        }
+
+        /// <summary>
+        /// Gets the amount of available space that can be written to
+        /// without forcing the underlying buffer to grow.
+        /// </summary>
+        public int FreeCapacity
+        {
+            get
+            {
+                Debug.Assert(this.rentedBuffer != null);
+                return this.rentedBuffer.Length - this.index;
+            }
+        }
+
+        /// <summary>
+        /// <inheritdoc/>
+        /// </summary>
+        /// <param name="count"></param>
+        public void Advance(int count)
+        {
+            Debug.Assert(this.rentedBuffer != null);
+            Debug.Assert(count >= 0);
+            Debug.Assert(this.index <= this.rentedBuffer.Length - count);
+            this.index += count;
+        }
+
+        /// <summary>
+        /// <inheritdoc/>
+        /// </summary>
+        /// <param name="sizeHint"></param>
+        /// <returns></returns>
+        public Memory<byte> GetMemory(int sizeHint = MinimumBufferSize)
+        {
+            CheckAndResizeBuffer(sizeHint);
+            return this.rentedBuffer.AsMemory(index);
+        }
+
+        /// <summary>
+        /// <inheritdoc/>
+        /// </summary>
+        /// <param name="sizeHint"></param>
+        /// <returns></returns>
+        public Span<byte> GetSpan(int sizeHint = MinimumBufferSize)
+        {
+            if (sizeHint == 0)
+            {
+                sizeHint = MinimumBufferSize;
+            }
+
+            CheckAndResizeBuffer(sizeHint);
+            return this.rentedBuffer.AsSpan(index);
+        }
+
+        /// <summary>
+        /// Clears the data written to the underlying buffer.
+        /// </summary>
+        public void Clear()
+        {
+            ClearHelper();
+        }
+
+        /// <summary>
+        /// Returns the rented memory back to the pool.
+        /// </summary>
+        public void Dispose()
+        {
+            if (this.rentedBuffer == null)
+            {
+                return;
+            }
+
+            ClearHelper();
+            byte[] toReturn = this.rentedBuffer;
+            this.rentedBuffer = null;
+            ArrayPool<byte>.Shared.Return(toReturn);
+        }
+
+        private void ClearHelper()
+        {
+            Debug.Assert(this.rentedBuffer != null);
+            Debug.Assert(this.index <= this.rentedBuffer.Length);
+
+            this.rentedBuffer.AsSpan(0, index).Clear();

```

Comment:
Curious if the following would be more efficient:
```
Array.Clear(this.rentedBuffer, 0, index);
````

Reply:
Good question. Both achieve the same thing, filling the segment with 0's. But peeking at the implementation, [`Array.Clear`](https://source.dot.net/#System.Private.CoreLib/src/System/Array.CoreCLR.cs,141) seems to do more work than [`Span<T>.Clear`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Span.cs,285). Both of them call the same `SpanHelpers.ClearWithoutReferences()` method (after checking whether the element type has references or not). But `Array.Clear` appears to do a lot more work before calling that method, including retrieving the method table and checking the array's dimentionality.